src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataReadmeControl.xaml.cs

```
@@ -0,0 +1,177 @@
+
+            _disposed = true;
+        }
+
+        private static void OnPropertyChanged(
+           DependencyObject dependencyObject,
+           DependencyPropertyChangedEventArgs e)
+        {
+            if (e.Property == PackageMetadataProperty && dependencyObject is PackageMetadataReadmeControl control)
+            {
+                control?.UpdateControl((DetailedPackageMetadata)e.OldValue, (DetailedPackageMetadata)e.NewValue);
+            }
+        }
+
+        private void UpdateControl(DetailedPackageMetadata oldValue, DetailedPackageMetadata newValue)
+        {
+            if (newValue is not null &&
+                (oldValue?.Id != newValue.Id ||
+                oldValue?.Version != newValue.Version))
+            {
+                var loadCts = new CancellationTokenSource();
+                var oldCts = Interlocked.Exchange(ref _loadCts, loadCts);
+                oldCts?.Cancel();
+                oldCts?.Dispose();
+
+
+                NuGetUIThreadHelper.JoinableTaskFactory
+                .RunAsync(async () =>
+                {
+                    await ReadmeViewModel.LoadReadmeAsync(newValue, _renderLocalReadme, _loadCts.Token);
+                })
+                .PostOnFailure(nameof(PackageMetadataReadmeControl));
+            }
+        }
+
+        private void PackageMetadataReadmeControl_DataContextChanged(object sender, DependencyPropertyChangedEventArgs e)
+        {
+            if (e.OldValue is ReadmePreviewViewModel oldViewModel && oldViewModel is not null)
+            {
+                oldViewModel.PropertyChanged -= ViewModel_PropertyChangedAsync;
+            }
+            ReadmeViewModel.PropertyChanged += ViewModel_PropertyChangedAsync;
+        }
+
+#pragma warning disable VSTHRD100 // Avoid async void methods
+        private async void ViewModel_PropertyChangedAsync(object sender, System.ComponentModel.PropertyChangedEventArgs e)
+#pragma warning restore VSTHRD100 // Avoid async void methods
```

It's best to keep the event handler synchronous and use JoinableTaskFactory.RunAsync to call an async method. An async event handler will swallow exceptions. Alternatively, we should handle exceptions in the async void method, but it's preferable to avoid async methods returning void if possible.

---

Okay so running a fire and forget RunAsync gives me a warning: https://github.com/Microsoft/VSSDK-Analyzers/blob/main/doc/VSSDK007.md

Running the property change synchronously could be an issue because it would block when rendering the readme, which I assume could be a problem if it takes a while to load because of the readme size. I think for now I'm going to scrap the property changed all together, since the only time the VM is updated is from the view. I know there's another comment about setting the model for the VM so this might change again